lang/LangPrimSource/PyrListPrim.cpp

```diff
@@ -492,41 +492,98 @@ int prEvent_Delta(struct VMGlobals *g, int numArgsPushed)
 	PyrSlot *a, key, dur, stretch, delta;
 	double fdur, fstretch;
 	int err;
+	PyrClass *restClass = s_rest->u.classobj;
+	PyrSlot *slot;
 
 	a = g->sp;  // dict
 
 	SetSymbol(&key, s_delta);
 	identDict_lookup(slotRawObject(a), &key, calcHash(&key), &delta);
 
 	if (NotNil(&delta)) {
-		slotCopy(a,&delta);
+		if (isKindOfSlot(&delta, restClass)) {
+			slot = slotRawObject(&delta)->slots;
+			err = slotDoubleVal(slot, &fdur);
+		} else {
+			err = slotDoubleVal(&delta, &fdur);
+		};
+		if (err) {
+			return err;
+		} else {
+			SetFloat(g->sp, fdur);
+			return errNone;
+		}
 	} else {
 		SetSymbol(&key, s_dur);
 		identDict_lookup(slotRawObject(a), &key, calcHash(&key), &dur);
-
 		err = slotDoubleVal(&dur, &fdur);
 		if (err) {
-			if (NotNil(&dur)) return err;
-			SetNil(a);
-			return errNone;
+			if (NotNil(&dur)) {
+				if (isKindOfSlot(&dur, restClass)) {
+					slot = slotRawObject(&dur)->slots;
+					err = slotDoubleVal(slot, &fdur);
+					if (err) return err;
+				} else {
+					return errWrongType;
+				}
+			} else {
+				SetNil(g->sp);
+				return errNone;
+			}
 		}
 
 		SetSymbol(&key, s_stretch);
 		identDict_lookup(slotRawObject(a), &key, calcHash(&key), &stretch);
 
 		err = slotDoubleVal(&stretch, &fstretch);
 		if (err) {
-			if (NotNil(&stretch)) return err;
-			SetFloat(a, fdur);
-			return errNone;
+			if (NotNil(&stretch)) {
+				if (isKindOfSlot(&stretch, restClass)) {
+					slot = slotRawObject(&stretch)->slots;
+					err = slotDoubleVal(slot, &fstretch);
+					if (err) return err;
+				} else {
+					return errWrongType;
+				}
+			} else {
+				SetFloat(g->sp, fdur);
+				return errNone;
+			}
 		}
 
-		SetFloat(a, fdur * fstretch );
+		SetFloat(g->sp, fdur * fstretch);
```

Comment:
For consistency, everything in this method should use `a` instead of `g->sp`; this makes it easier to understand that these modifications are being made to the same object in the stack.